### ISSUE TITLE
Publish rust-hoku crates

### ISSUE THREAD

**Issue by sanderpick** [recallnet/rust-recall#17]
DOD:
- [x] `rust-hoku` packages are available on crates.io
  - [ ] we are doing git tag based releases
- [ ] `basin-s3` packages are available on crates.io
  - [ ] we are doing git tag based releases
- [x] our version of `ipc` packages are available on crates.io (note: we should namespace with hoku)
  - [ ] we are doing git tag based releases
- [x] https://github.com/consensus-shipyard/ipc/issues/1000 is completed

**Comment by sanderpick** [comment]
@joewagner can you take another pass at writing down all the crate names we are going to publish? ie, edit this issue's description above.

here's my understanding of how the publishing works:
- crate.io doesn't have orgs, just individual accounts
- somebody will have to first manually publish each crate, and then we can use `cargo owner add` to add a GitHub team so that we can have more than one individual account owner. i can do that from a fresh cold account. 
- the cold account will need to generate an API key that we put in org wide GitHub secrets (https://doc.rust-lang.org/cargo/reference/config.html?highlight=CARGO_REGISTRY_TOKEN#credentials), which can be used in your flow

sound right?

**Comment by joewagner** [comment]
> [@joewagner](https://github.com/joewagner) can you take another pass at writing down all the crate names we are going to publish? ie, edit this issue's description above.

Absolutely, I'll build a graph that shows the deps we will want to publish on Crates.

> here's my understanding of how the publishing works:
> 
> * crate.io doesn't have orgs, just individual accounts
> * somebody will have to first manually publish each crate, and then we can use `cargo owner add` to add a GitHub team so that we can have more than one individual account owner. i can do that from a fresh cold account.
> * the cold account will need to generate an API key that we put in org wide GitHub secrets (https://doc.rust-lang.org/cargo/reference/config.html?highlight=CARGO_REGISTRY_TOKEN#credentials), which can be used in your flow
> 
> sound right?

This plan sounds great.  We might be able to do all this without first manually publishing, but if that makes the most sense I don't see any problem with it.

**Comment by joewagner** [comment]
@sanderpick here is a heavily consolidated dependency graph with the goal of understanding what's required to publish the rust-recall sdk and cli

```
// rust-recall repo
rust_recall (vitual workspace)
	recall_cli -- |
		recall_provider
			// in ipc repo
			fendermint_actor_blobs_shared
			fendermint_eth_api
			fendermint_vm_actor_interface
			fendermint_vm_message
			ipc-api

		recall_sdk
			// in ipc repo
			fendermint_actor_blobs_shared
			fendermint_actor_bucket
			fendermint_actor_recall_config_shared
			fendermint_actor_machine
			fendermint_actor_timehub
			fendermint_vm_actor_interface
			ipc_actors_abis
			ipc-api
			// in this repo
			recall_provider
			recall_signer

		recall_signer
			// in ipc repo
			fendermint_crypto
			fendermint_vm_actor_interface
			tendermint-rpc
			ipc-api
			// in this repo
			recall_provider

// ipc repo
ipc (virtual)
	fendermint_app --|
                // the entaglement (`storage` and `entagler`) repo is the only dep that we need to handle publishing for ipc
		storage
		entangler --|
			storage		

//  entanglement repo
entanglement (virtual)
	entaglement_cli -- |
			entangler
			storage
        entangler
        storage
```

Regarding `basin-s3`: it's only dependencies are `recall_sdk`, `recall_provider`, and `recall_signer`.  So, it would be the last crate we need to publish.

As you mention, we will want to rename several of these packages so they make sense when published on `crates.io`.
My suggestions are below, but if you have ideas I don't have any real preference.

ipc -> recall_ipc
entanglement_cli -> recall_entaglement_cli
entangler -> recall_entangler
storage -> recall_entaglement_storage
basin_s3 -> recall_s3  (not sure on this one, do we want to change basin to recall here?)

Lastly, and maybe most contentious...  The Cargo and Crates ecosystem seems to be split on if crate names should be snake_case or kebob-case.  Do we prefer snake?

**Comment by joewagner** [comment]
TL;DR;
we should publish entanglement, then ipc, then rust-recall, then basin-s3 (assuming we need to)

**Comment by sanderpick** [comment]
looks good! shoot, so do we have to rename all the internal IPC crates? ie, `fendermint_eth_api` -> `recall_fendermint_eth_api`? if so, we could consider not publishing IPC at all and instead create a standalone crate that contains all our types or something... I think David suggested something like this a while back. but maybe that's more cumbersome than just renaming the internal IPC crates. 

basin-s3 is already renamed: https://github.com/recallnet/recall-s3

i like the hyphens but no strong pref. iroh uses hyphens... happy to adopt their opinion.

**Comment by sanderpick** [comment]
actually, lets just use underscores... otherwise we'd have to change everything. ie, https://github.com/recallnet/rust-recall/blob/main/sdk/Cargo.toml#L2

**Comment by joewagner** [comment]
I agree hyphens are nicer, but changing will be a bit of a pain.  Underscores sound great, and we can always change later on if it seems appropriate.

I can't say for sure since I'm pretty new to rust packages.  I _think_ to publish a package to crates, it's entire dependency tree needs to also be available on crates. So this means all the workspace members in `ipc` need to be renamed, unless consensus-shipyard/ipc is going to publish a version that matches what we have in our repo (which might be the case for some of the packages?).

**Comment by joewagner** [comment]
@sanderpick How should we proceed here?  I am happy to do all of this work, but it seems like you might have to do some of it since you are probably going to be the one to own the official Crates.io account?

**Comment by joewagner** [comment]
> i don't need to own it... i will have to create a cold email account, but can put the creds in 1pass where you can find them. i guess we better try changing all the internal IPC package names and do some test publishing to see what happens.
> 
> i'll setup the cold account this evening.

Nice!  Sounds great.  I'll start working on all this, ping me when the credentials are in 1pass.

**Comment by sanderpick** [comment]
how about...

ipc -> recall_ipc
entanglement_cli -> recall_entangler_cli
entangler -> recall_entangler
storage -> recall_entangler_storage
recall_s3

**Comment by joewagner** [comment]
> how about...
> 
> ipc -> recall_ipc entanglement_cli -> recall_entangler_cli entangler -> recall_entangler storage -> recall_entangler_storage recall_s3

Works for me!  I don't know how crates works, but I'll republish and see about removing the incorrect names.
I'll have to make a quick PR to https://github.com/recallnet/entanglement as well

**Comment by stbrody** [comment]
Blocked on code review from @sanderpick

**Comment by sanderpick** [comment]
@joewagner can you remind me, what PRs are you waiting on me for?

**Comment by joewagner** [comment]
> [@joewagner](https://github.com/joewagner) can you remind me, what PRs are you waiting on me for?

This one, and the others listed in that PR https://github.com/recallnet/ipc/pull/580

**Comment by joewagner** [comment]
Let me know if you want to chat about anything.  There's a lot changes and it's potentially a little confusing.  If you are ok with everything, I'll have to do a bunch of manual rebase conflict resolution.   After they PRs would need a proper approval.

**Comment by sanderpick** [comment]
Since we're not in a huge hurry to get this out anymore, I wonder if we should wait until we merge upstream with all the audit fixes before doing this because these changes will make that harder. wdyt?

**Comment by joewagner** [comment]
> Since we're not in a huge hurry to get this out anymore, I wonder if we should wait until we merge upstream with all the audit fixes before doing this because these changes will make that harder. wdyt?

The upstream merges getting more complex is a definite downside to this. 
I haven't ever tried doing an upstream merge for this repo, so I'm not sure how much of a pain it is.  With that said I'm leaning toward merging this soon, since it should make any changes to the sdk more practical to get out to the public.
But... I'm definitely ok waiting if that makes things easier overall.
